File: BrickKit/app/src/main/java/com/wayfair/brickkitdemo/MainActivityFragment.java
/*
 * Copyright © 2017-2020 Wayfair. All rights reserved.
 */
package com.wayfair.brickkitdemo;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;

import com.wayfair.brickkit.brick.BaseBrick;
import com.wayfair.brickkit.BrickFragment;
import com.wayfair.brickkit.padding.BrickPadding;
import com.wayfair.brickkit.padding.BrickPaddingFactory;
import com.wayfair.brickkit.size.BrickSize;
import com.wayfair.brickkit.size.PercentageBrickSize;
import com.wayfair.brickkitdemo.bricks.UnusedBrick;
import com.wayfair.brickkitdemo.bricks.UsedBrick;

import java.util.ArrayList;

/**
 * A placeholder fragment containing a simple view.
 */
public class MainActivityFragment extends BrickFragment {
    private static final BrickSize ONE_FIFTH = new PercentageBrickSize(.2f);
    private static final BrickSize TWO_FIFTH = new PercentageBrickSize(.4f);

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        BrickPadding padding = new BrickPaddingFactory(getResources()).getInnerOuterBrickPadding(R.dimen.four_dp, R.dimen.no_dp);

        ArrayList<BaseBrick> usedBricks = new ArrayList<>();

        final UsedBrick usedBrick = new UsedBrick(
                TWO_FIFTH,
                padding,
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        getFragmentManager().beginTransaction()
                                .replace(R.id.content, new SimpleBrickFragment())
                                .addToBackStack(null)
                                .commit();
                    }
                }
        );
        usedBricks.add(usedBrick);

        // Fake waiting as in a API request to show off placeholder when data is not ready
        new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
            @Override
            public void run() {
                usedBrick.setText("Simple Brick View");
                dataManager.refreshItem(usedBrick);
            }
        }, 2000);
        usedBricks.add(
                new UsedBrick(
                        TWO_FIFTH,
                        padding,
                        "Add/Remove Brick View",
                        new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                getFragmentManager().beginTransaction()
                                        .replace(R.id.content, new AddRemoveBrickFragment())
                                        .addToBackStack(null)
                                        .commit();
                            }
                        }
                )
        );
        usedBricks.add(
                new UsedBrick(
                        TWO_FIFTH,
                        padding,
                        "Infinite Scroll Brick View",
                        new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                getFragmentManager().beginTransaction()
                                        .replace(R.id.content, new InfiniteScrollBrickFragment())
                                        .addToBackStack(null)
                                        .commit();
                            }
                        }
                )
        );
        usedBricks.add(
                new UsedBrick(
                        TWO_FIFTH,
                        padding,
                        "Staggered Infinite Scroll Brick View",
                        new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                getFragmentManager().beginTransaction()
                                        .replace(R.id.content, new StaggeredInfiniteScrollBrickFragment())
                                        .addToBackStack(null)
                                        .commit();
                            }
                        }
                )
        );
        usedBricks.add(
                new UsedBrick(
                        TWO_FIFTH,
                        padding,
                        "Fragment Brick View",
                        new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                getFragmentManager().beginTransaction()
                                        .replace(R.id.content, new FragmentBrickFragment())
                                        .addToBackStack(null)
                                        .commit();
                            }
                        }
                )
        );

        for (int i = 0; i < usedBricks.size() + 2; i++) {
            BrickSize first;
            BrickSize last;

            if (i % 2 == 0) {
                first = ONE_FIFTH;
                last = TWO_FIFTH;
            } else {
                first = TWO_FIFTH;
                last = ONE_FIFTH;
            }

            UnusedBrick unusedBrick1 = new UnusedBrick(first, padding);
            dataManager.addLast(unusedBrick1);

            if (i == 0 || i == usedBricks.size() + 1) {
                UnusedBrick unusedBrick = new UnusedBrick(
                        TWO_FIFTH,
                        padding
                );
                dataManager.addLast(unusedBrick);
            } else {
                dataManager.addLast(usedBricks.get(i - 1));
            }

            UnusedBrick unusedBrick2 = new UnusedBrick(last, padding);
            dataManager.addLast(unusedBrick2);

        }

    }
}

File: BrickKit/app/src/main/java/com/wayfair/brickkitdemo/mainfragment/MainActivityFragment.kt
/*
 * Copyright © 2017-2021 Wayfair. All rights reserved.
 */
package com.wayfair.brickkitdemo.mainfragment

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import androidx.fragment.app.Fragment
import com.wayfair.brickkit.brick.BaseBrick
import com.wayfair.brickkit.size.BrickSize
import com.wayfair.brickkit.size.PercentageBrickSize
import com.wayfair.brickkitdemo.AddRemoveBrickFragment
import com.wayfair.brickkitdemo.BrickFragment
import com.wayfair.brickkitdemo.infinitescroll.InfiniteScrollBrickFragment
import com.wayfair.brickkitdemo.R
import com.wayfair.brickkitdemo.simplebrick.SimpleBrickFragment
import com.wayfair.brickkitdemo.StaggeredInfiniteScrollBrickFragment
import com.wayfair.brickkitdemo.fragmentbrick.FragmentBrickFragment
import com.wayfair.brickkitdemo.mainfragment.bricks.ActiveBrick
import com.wayfair.brickkitdemo.mainfragment.bricks.PassiveBrick
import java.util.concurrent.TimeUnit

/**
 * Main fragment that links to the other child fragments. Additionally this shows an example of placeholders.
 */
class MainActivityFragment : BrickFragment() {
    private var rowNumber = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val placeholderBrick = ActiveBrick(resources, R.string.empty) { changeFragment(SimpleBrickFragment.newInstance()) }
            .also {
                // Fake waiting as in a API request to show off placeholder when data is not ready
                Handler(Looper.getMainLooper()).postDelayed(
                    { it.text = getString(R.string.simple_brick_title) },
                    TimeUnit.SECONDS.toMillis(2)
                )
            }

        addRow(PassiveBrick(TWO_FIFTH, resources))
        addRow(placeholderBrick)
        addRow(ActiveBrick(resources, R.string.add_remove_brick_title) { changeFragment(AddRemoveBrickFragment()) })
        addRow(ActiveBrick(resources, R.string.infinite_scroll_brick_title) { changeFragment(InfiniteScrollBrickFragment()) })
        addRow(ActiveBrick(resources, R.string.staggered_infinite_scroll_brick_title) { changeFragment(StaggeredInfiniteScrollBrickFragment()) })
        addRow(ActiveBrick(resources, R.string.fragment_brick_title) { changeFragment(FragmentBrickFragment()) })
        addRow(PassiveBrick(TWO_FIFTH, resources))
    }

    private fun addRow(middleBrick: BaseBrick) {
        dataManager.addLast(PassiveBrick(if (rowNumber % 2 == 0) ONE_FIFTH else TWO_FIFTH, resources))
        dataManager.addLast(middleBrick)
        dataManager.addLast(PassiveBrick(if (rowNumber % 2 == 0) TWO_FIFTH else ONE_FIFTH, resources))
        rowNumber += 1
    }

    private fun changeFragment(fragment: Fragment) {
        fragmentManager?.beginTransaction()?.replace(R.id.content, fragment)?.addToBackStack(null)?.commit()
    }

    companion object {
        private val ONE_FIFTH: BrickSize = PercentageBrickSize(.2f)
        private val TWO_FIFTH: BrickSize = PercentageBrickSize(.4f)
    }
}
